core/services/ocr2/plugins/ocr2keeper/coordinator.go

```diff
@@ -0,0 +1,234 @@
+				return
+			}
+		}(chPerformLogs)
+	}
+}
+
+type LogCoordinator struct {
+	utils.StartStopOnce
+	logger          logger.Logger
+	logPoller       logpoller.LogPoller
+	registryAddress common.Address
+	lookbackBlocks  int64
+	provider        *performLogProvider
+	registry        *registry.KeeperRegistry
+	stop            chan struct{}
+}
+
+func NewLogCoordinator(
+	logger logger.Logger,
+	logPoller logpoller.LogPoller,
+	registryAddress common.Address,
+	client evmclient.Client,
+	lookbackBlocks int64,
+) (*LogCoordinator, error) {
+	var err error
+
+	contract, err := registry.NewKeeperRegistry(common.HexToAddress("0x"), client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add log filters for the log poller so that it can poll and find the logs that
+	// we need.
+	err = logPoller.MergeFilter([]common.Hash{
+		registry.KeeperRegistryUpkeepPerformed{}.Topic(),
+	}, []common.Address{registryAddress})
+	if err != nil {
+		return nil, err
+	}
+
+	return &LogCoordinator{
+		logger:          logger,
+		logPoller:       logPoller,
+		registryAddress: registryAddress,
+		lookbackBlocks:  lookbackBlocks,
+		provider:        &performLogProvider{subscriptions: make(map[string]chan plugintypes.PerformLog)},
+		registry:        contract,
+		stop:            make(chan struct{}),
+	}, nil
+}
+
+func (c *LogCoordinator) PerformLogProvider() plugintypes.PerformLogProvider {
+	return c.provider
+}
+
+func (c *LogCoordinator) run() {
+	defer func() {
+		// gracefully handle panics by printing the stack trace and restarting
+		// the process after some wait time
+		if err := recover(); err != nil {
+			c.logger.Errorf("log coordinator run routine had a panic: %s", err)
+
+			// print the stack trace
+			debug.PrintStack()
+
+			// restart the process after a wait time
+			<-time.After(1 * time.Second)
+
+			c.logger.Errorf("log coordinator run routine is restarting")
+			go c.run()
+		}
+	}()
+
+	cadence := time.Second
+	t := time.NewTimer(cadence)
+	for {
+		select {
+		case <-t.C:
+			ctx, cancel := context.WithTimeout(context.Background(), cadence)
+
+			end, err := c.logPoller.LatestBlock(pg.WithParentCtx(ctx))
+			if err != nil {
+				c.logger.Errorf("failed to get latest block from log poller: %s", err)
+				cancel()
+				continue
+			}
+
+			logs, err := c.logPoller.LogsWithSigs(
+				end-c.lookbackBlocks,
+				end,
+				[]common.Hash{
+					registry.KeeperRegistryTransmitted{}.Topic(),
+				},
+				c.registryAddress,
+				pg.WithParentCtx(ctx),
+			)
+			if err != nil {
+				c.logger.Errorf("failed to collect logs from log poller: %s", err)
+				cancel()
+				continue
+			}
+			cancel()
+
+			performed, err := c.unmarshalLogs(logs)
+			if err != nil {
+				c.logger.Errorf("failed to unmarshal logs: %s", err)
+				cancel()
+				continue
+			}
+
+			for _, p := range performed {
+				// broadcast log to subscribers
+				c.provider.Broadcast(plugintypes.PerformLog{
+					Key: plugintypes.UpkeepKey(fmt.Sprintf("%d|%s", p.Raw.BlockNumber, p.Id.String())),
```

Comment:
I think this will need to be the checkBlockNumber in log as you expect that in the plugin